### ISSUE TITLE
DTLS server does not respond to retransmitted "Client Hello"

### ISSUE THREAD

**Issue by mniestroj** [openssl/openssl#18635]
Version: 20af01d46ca4086f6b1339b67c15c81d8b4a040d (4 commits before `OpenSSL_1_1_1p`)

In order to reproduce the issue, use https://github.com/mniestroj/openssl/tree/reproduce-dtls-handshake-deadlock or just apply https://github.com/mniestroj/openssl/commit/b314ee46815c10d52c94f1f13afd67035edcb91e commit, which simulates dropping reply to first "Client Hello" packet (this means that client does not receive response to "Client Hello" and continues retransmitting it).

```
$ apps/openssl s_server -dtls1_2 -accept 192.0.2.2:5684 -psk_identity mark-one-id@one -psk 3172306e6d406e -cipher PSK-AES128-CCM8 -nocert -security_debug_verbose
Using default temp DH parameters
ACCEPT

Security callback: Version=???: yes
Security callback: Version=???: yes
comparing msg_hdr.seq 0 s->d1->handshake_read_seq 0 (s->statem.hand_state 0)
Don't send this message!
comparing msg_hdr.seq 0 s->d1->handshake_read_seq 1 (s->statem.hand_state 21)
comparing msg_hdr.seq 0 s->d1->handshake_read_seq 1 (s->statem.hand_state 21)
comparing msg_hdr.seq 0 s->d1->handshake_read_seq 1 (s->statem.hand_state 21)
comparing msg_hdr.seq 0 s->d1->handshake_read_seq 1 (s->statem.hand_state 21)
comparing msg_hdr.seq 0 s->d1->handshake_read_seq 1 (s->statem.hand_state 21)
comparing msg_hdr.seq 0 s->d1->handshake_read_seq 1 (s->statem.hand_state 21)
comparing msg_hdr.seq 0 s->d1->handshake_read_seq 1 (s->statem.hand_state 21)
comparing msg_hdr.seq 0 s->d1->handshake_read_seq 1 (s->statem.hand_state 21)
comparing msg_hdr.seq 0 s->d1->handshake_read_seq 1 (s->statem.hand_state 21)
comparing msg_hdr.seq 0 s->d1->handshake_read_seq 1 (s->statem.hand_state 21)
comparing msg_hdr.seq 0 s->d1->handshake_read_seq 1 (s->statem.hand_state 21)
comparing msg_hdr.seq 0 s->d1->handshake_read_seq 1 (s->statem.hand_state 21)
comparing msg_hdr.seq 0 s->d1->handshake_read_seq 1 (s->statem.hand_state 21)
```

Recurring `comparing msg_hdr.seq 0 ...` messages are printed whenever new retransmitted "Client Hello" message is received. Unfortunately openssl treats that message as invalid. It looks like handling of retransmitted "Client Hello" is not handled correctly.

mbedTLS based DTLS client was used, though it should be reproducible with any DTLS client.

See that "Client Hello" messages are visible in Wireshark as well:
![wireshak_openssl](https://user-images.githubusercontent.com/10817297/175097565-85dcb12a-fb6e-4b44-be63-f0f055a14cb9.png)

This is quite severe bug, because restarting client (or starting new handshake) does not solve the issue and this means that there is persistent deadlock for such client. A workaround was developed, which seems to solve the issue: https://github.com/mniestroj/openssl/commit/18ad4353e8b5f35d05e742097e416b8d6a0d3f6d. After applying this workaround, DTLS server properly continues:

```
$ apps/openssl s_server -dtls1_2 -accept 192.0.2.2:5684 -psk_identity mark-one-id@one -psk 3172306e6d406e -cipher PSK-AES128-CCM8 -nocert -security_debug_verbose
Using default temp DH parameters
ACCEPT
Security callback: Version=???: yes
Security callback: Version=???: yes
comparing msg_hdr.seq 0 s->d1->handshake_read_seq 0 (s->statem.hand_state 0)
Don't send this message!
comparing msg_hdr.seq 0 s->d1->handshake_read_seq 1 (s->statem.hand_state 21)
special case for handling Client Hello retransmit 1 21
comparing msg_hdr.seq 1 s->d1->handshake_read_seq 1 (s->statem.hand_state 21)
Security callback: Version=???: yes
Security callback: : yes
Security callback: Shared Signature Algorithm scheme=ecdsa_secp521r1_sha512, security bits=256: yes
Security callback: Shared Signature Algorithm scheme=rsa_pkcs1_sha512, security bits=256: yes
Security callback: Shared Signature Algorithm scheme=ecdsa_secp384r1_sha384, security bits=192: yes
Security callback: Shared Signature Algorithm scheme=rsa_pkcs1_sha384, security bits=192: yes
Security callback: Shared Signature Algorithm scheme=ecdsa_secp256r1_sha256, security bits=128: yes
Security callback: Shared Signature Algorithm scheme=rsa_pkcs1_sha256, security bits=128: yes
Security callback: Shared Signature Algorithm scheme=ecdsa_secp521r1_sha512, security bits=256: yes
Security callback: Shared Signature Algorithm scheme=rsa_pkcs1_sha512, security bits=256: yes
Security callback: Shared Signature Algorithm scheme=ecdsa_secp384r1_sha384, security bits=192: yes
Security callback: Shared Signature Algorithm scheme=rsa_pkcs1_sha384, security bits=192: yes
Security callback: Shared Signature Algorithm scheme=ecdsa_secp256r1_sha256, security bits=128: yes
Security callback: Shared Signature Algorithm scheme=rsa_pkcs1_sha256, security bits=128: yes
Security callback: Shared Ciphersuite=PSK-AES128-CCM8, security bits=128: yes
comparing msg_hdr.seq 2 s->d1->handshake_read_seq 2 (s->statem.hand_state 26)
comparing msg_hdr.seq 3 s->d1->handshake_read_seq 3 (s->statem.hand_state 31)
-----BEGIN SSL SESSION PARAMETERS-----
...
```

**Comment by t8m** [comment]
@bernd-edlinger was recently looking at DTLS so perhaps he could comment?

**Comment by bernd-edlinger** [comment]
That is a limitation of the s_server command.
It does only allow one single connection at a time.
Try to use the -timeout option as a workaround,
to allow the s_server to disconnect after a timeout
is detected.
'openssl s_server -dtls -timeout`
should work.

**Comment by mattcaswell** [comment]
> That is a limitation of the s_server command.

No. There is a real bug here. I have it replicated locally. Still looking at it.

**Comment by mattcaswell** [comment]
Fix for this in #18654.

Note that this is only a problem when using "stateful" cookie exchange. The preferred way is to use "stateless" cookies via the `DTLSv1_listen` API - and that API does not suffer from this problem. `s_server` implements this via the `-listen` flag. See:

https://www.openssl.org/docs/man3.0/man3/DTLSv1_listen.html
